src/components/Topology/Topology.js

```diff
@@ -240,25 +320,158 @@ export default ({ modelData, width, height }) => {
+        const iconX = d3.event.x - radius;
+        const iconY = d3.event.y - radius;
+
+        return `translate(${iconX}, ${iconY})`;
+      });
+      updateRelations(relationLine);
+    }
 
-    relationLine
-      .classed("relation", true)
-      .append("line")
-      .attr("x1", (d) => getRelationPosition(d).x1)
-      .attr("y1", (d) => getRelationPosition(d).y1)
-      .attr("x2", (d) => getRelationPosition(d).x2)
-      .attr("y2", (d) => getRelationPosition(d).y2)
-      .attr("stroke", "#666666")
-      .attr("stroke-width", 2);
+    function dragended() {
+      const app = d3.select(this);
+      app.select("circle").attr("stroke", "#888888");
+      // const name = app.attr("data-name");
+      // const annotations = getRect(app);
+      // const updateResponse = updateAnnotations(conn, name, annotations);
+      // updateResponse.then((data) => {
+      //   if (!data) {
+      //     setFlashMessage({
+      //       type: "error",
+      //       message: "Sorry - the position of your apps could not be updated.",
+      //     });
+      //   } else {
+      //     setFlashMessage({
+      //       type: "success",
+      //       message: "The position of your apps have been updated.",
+      //     });
+      //   }
+      // });
+    }
+
+    if (!isReadOnly) {
+      appIcon.call(
+        d3.drag().on("start", dragstarted).on("drag", drag).on("end", dragended)
+      );
+    }
+
+    setRelations(relationLine);
 
     appIcons.exit().remove();
     relationLines.exit().remove();
 
     return () => {
       topo.remove();
     };
-  }, [applications, deltaX, deltaY, height, width, maxX, maxY, relations]);
-  return <svg ref={ref} />;
+  }, [
+    applications,
+    conn,
+    deltaX,
+    deltaY,
+    isFullscreen,
+    isReadOnly,
+    maxX,
+    maxY,
+    relations,
+  ]);
+
+  // XXX Remove flash messaging until we can save results back to juju
+  // Reset flash messages after 3 seconds
+  // useEffect(() => {
+  //   if (flashMessage.message) {
+  //     window.clearFlashMessage = setTimeout(() => {
+  //       setFlashMessage({});
+  //     }, 3000);
+  //   }
+  //   return () => {
+  //     clearTimeout(window.clearFlashMessage);
+  //   };
+  // }, [flashMessage]);
+
+  // Close topology, if open, on Escape key press
+  useEffect(() => {
+    const closeOnEscape = function (e) {
+      if (e.code === "Escape" && isFullscreen) {
+        setIsFullscreen(false);
+      }
+    };
+    document.addEventListener("keydown", closeOnEscape);
+    return () => {
+      document.removeEventListener("keydown", closeOnEscape);
+    };
+  });
+
+  console.log("texts");
+  if (Object.entries(modelData?.applications || {}).length) {
+    return (
+      <div
+        className="topology"
+        data-fullscreen={isFullscreen}
+        data-readonly={isReadOnly}
+        ref={topologyRef}
+      >
+        <div className="topology__inner">
+          {/* Remove flash messaging until we can save positions back to juju */}
```

Comment:
Lets remove this until we can implement it as this might not be the place we ultimately decide to put it.